src/basic/copy.c

```diff
@@ -525,7 +526,8 @@ static int fd_copy_directory(
         r = mkdirat(dt, to, st->st_mode & 07777);
         if (r >= 0)
                 created = true;
-        else if (errno == EEXIST && (copy_flags & COPY_MERGE))
+        else if (errno == EEXIST &&
+                 (copy_flags & COPY_MERGE || (copy_flags & COPY_MERGE_EMPTY && dir_is_empty_at(dt, to))))
```

Comment:
this would misunderstand a failure in "dir_is_empty_at()" as "yes it's an empty dir"... please add proper error handling and propagation to the dir_is_empty_at()...

Reply:
hmm, also: please do the dir_is_empty_at() thing first if the new bit is set, and only if that returns ENOENT continue with the mkdirat(). Why? creating a directory might fail with EROFS or EPERM or so even if it already exists, if for some reason the caller has no privs to create a directory (think selinux, read-only file systems, overlayfs, …). Let's hence sidestep the issue and use dir_is_empty_at() first (which is more likely to succeed in such a case), and then use mkdirat() only if we detected the dir doesn't exist yet.